Add bundle contents API

kashizui opened the pull request:
And update the CLI to use the new API.
- [x] wait (Friday 5/6)
- [x] download (Saturday 5/7)
- [x] cat (Saturday 5/7)
- [x] info, with contents (Sunday 5/8)
- [x] worksheet_util.interpret_file_genpath
- [x] upload
- [x] optimize and clean up upload 
- [x] fix CI tests
- [x] squash and rebase

Fixes #335, #495

kashizui committed:
Add bundle contents API

kashizui commented:
Still needs to be rebased on #433

klopyrev commented on codalab/rest/bundle.py:
```diff
@@ -33,21 +567,34 @@ def get_blob(uuid, path=''):
+            start, end = byte_range
+            fileobj = local.download_manager.read_file_section(uuid, path, start, end - start + 1, gzipped)
+        elif head_lines or tail_lines:
+            fileobj = local.download_manager.summarize_file(uuid, path, head_lines, tail_lines, 128, None, gzipped)
```
Don't hard code the maximum line length.

klopyrev replied:
You could give it a default, though. Although, this is minor.

klopyrev commented:
Generally, looks pretty good. Don't forget that you also need to change worksheet_util.interpret_file_genpath, since that makes some calls to get_target_info and cat_target.

klopyrev commented:
I only looking at the downloading stuff, not the other bundle stuff, though.

kashizui commented:
@percyliang Finished with upload! I still need to fix the CircleCI test, but everything else is working and ready to be reviewed.

(The issue with the CircleCI test is that the RESTified commands, in particular `cl upload`, no longer work with the `local::` CodaLab instance, as we're now phasing out local instances. But this makes the `wadd` test a bit harder to implement, since we have to run multiple instances of the REST server. I'll give it a shot though soon.)

kashizui committed:
Refactor 'copy' test

 - Allow user to specify instance to log out from in `logout`
 - Create full servers in copy test instead of 'local' instance

kashizui commented:
@percyliang Fixed the test! All ready for review.
Required some major refactoring, which might be easier to pick apart by viewing that commit alone.

kashizui commented:
@percyliang I'm seeing a 40% improvement in speed for uploading a ~70mb directory on wstest, and a whopping 12x improvement in download speed. I've deployed this branch to wstest, so you can check out this branch and try it out yourself.

percyliang commented:
@kashizui : great news about the download speed; upload speed is still worrisome though.  Konstantin was getting ~60 MB/sec, whereas even on localhost, I'm only getting ~6 MB/sec.  (I'd test on localhost, not wstest, to make sure Internet speed is not the bottleneck factor.)

kashizui commented:
Ah ok I was wondering what numbers Konstantin had gotten. I'll look into
this more later this week..
On Wed, Jul 6, 2016 at 10:55 Percy Liang notifications@github.com wrote:

> @kashizui https://github.com/kashizui : great news about the download
> speed; upload speed is still worrisome though. Konstantin was getting ~60
> MB/sec, whereas even on localhost, I'm only getting ~6 MB/sec. (I'd test on
> localhost, not wstest, to make sure Internet speed is not the bottleneck
> factor.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codalab/codalab-cli/pull/434#issuecomment-230852644,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAlhta_1H9cfiAgY9qnpJPE0q7Q3B0RZks5qS-v0gaJpZM4Ic11h
> .

percyliang commented on codalab/client/json_api_client.py:
```diff
@@ -468,3 +470,61 @@ def update_authenticated_user(self, data, params=None):
+    @wrap_exception('Unable to upload contents of bundle {1}')
+    def upload_contents_blob(self, bundle_id, fileobj=None, params=None,
+                             progress_callback=None):
+        request_path = '/bundles/%s/contents/blob/' % bundle_id
```
Please document the arguments here, especially since the uploading logic has so many cases.

kashizui replied:
I will document the arguments here, but more importantly the query parameters are already documented in `rest/bundle.py`.

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -728,15 +732,24 @@ def do_status_command(self, args):
-        help='Logout of the current session.',
+        help='Logout of the current session, or a specific instance.',
+        arguments=(
+            Commands.Argument('alias', help='Alias or URL of instance from which to logout. Default is the current session.', nargs='?'),
```
name of argument would make more sense as 'instance' rather than 'alias'

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -801,61 +814,71 @@ def do_upload_command(self, args):
+            'worksheet': worksheet_uuid
+        })
+
+        if args.contents is not None:
```
Document more explicitly the ways people can upload by numbering them
1) Contents
2) URL
3) Local paths (usually one)

percyliang commented on codalab/lib/print_util.py:
```diff
@@ -12,3 +15,45 @@ def clear_line(f=sys.stderr):
+        else:
+            self.f.write(formatting.ratio_str(formatting.size_str, bytes_done, bytes_total))
+        speed = float(bytes_done) / (time.time() - self.start_time)
+        self.f.write(' [%sbps]' % formatting.size_str(speed))
```
This prints out 'mbps' which is not what we want - I'd go with MB/sec explicitly (don't need formatting.size_str, since the dynamic range here isn't going to be that big)

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -886,16 +909,20 @@ def do_download_command(self, args):
+        if target_info['type'] == 'link':
             raise UsageError('Downloading symlinks is not allowed.')
 
+        # TODO(sckoo): print progress
```
Is this easy to just do in this PR?

kashizui replied:
I thought it would be hard but I just did it :)

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -1542,11 +1568,13 @@ def print_target_info(self, client, target, decorate, maxlines=10, fail_if_not_e
             else:
-                client.cat_target(target, self.stdout)
+                with closing(client.fetch_contents_blob(target[0], target[1])) as contents:
+                    self.stdout.write(contents.read())
```
We should stream the contents through using copyfileobj...otherwise if you cat a large file, it waits for everything to be downloaded before printing anything out.

On a related note, we should support 'head' and 'tail' in addition to 'cat'.  Add as another PR.

percyliang commented on codalab/lib/zip_util.py:
```diff
@@ -85,3 +91,78 @@ def unpack(ext, source, dest_path):
+    """
+    Create a single flat tarfile containing all the sources.
+
+    Caller is responsible for closing the returned file-like object.
```
say fileobj explicitly

percyliang commented on codalab/lib/zip_util.py:
```diff
@@ -85,3 +91,78 @@ def unpack(ext, source, dest_path):
+
+    sources = map(resolve_source, sources)
+
+    # For efficiency, return singular files and directories directly
```
singular ==> single

percyliang commented on codalab/worker/bundle_manager.py:
```diff
@@ -111,6 +112,10 @@ def _stage_bundles(self):
         bundles_to_stage = []
         for bundle in bundles:
+            # Ignore non-derived (i.e. executable) bundles
+            if not isinstance(bundle, DerivedBundle):
```
non-derived is not-executable (comment is ambiguous)

Do we need this statement?  Edit: I understand why now, but will not be necessary if we have an 'uploading' state.

percyliang commented on test-cli.py:
```diff
@@ -742,65 +769,93 @@ def stress(use_time, request_time, use_memory, request_memory, use_disk, request
+        os.environ['CODALAB_USERNAME'] = test_username
+        os.environ['CODALAB_PASSWORD'] = test_password
+        subprocess.call('cl work ' + remote_worksheet, shell=True)
+        subprocess.call('cl gls > /dev/null', shell=True)
```
Why this command?

kashizui replied:
Just the first rest-ified command I could think of. I'll give a shot at fixing the double-login thing now, but if it doesn't go well I'll make sure to document this line better.

percyliang commented on test-cli.py:
```diff
@@ -742,65 +769,93 @@ def stress(use_time, request_time, use_memory, request_memory, use_disk, request
 def test(ctx):
     '''Test copying between instances.'''
-    # Figure out the current instance
+    # Figure out the current instance and use it as the 'remote' instance
```
Cool, testing on remote is good if we're going to deprecate local::

percyliang commented on test-cli.py:
```diff
@@ -742,65 +769,93 @@ def stress(use_time, request_time, use_memory, request_memory, use_disk, request
+
+    # TODO(sckoo): refactor test-cli to allow each module to define cleanup method
+    try:
+        wait_until_substring(server_proc.stdout, 'BundleRPCServer serving')
```
Write a comment to get rid of BundleRPCServer later

percyliang commented:
cl upload with multiple files doesn't work, I get:
  File "/home/pliang/research/codalab/codalab-cli/codalab/lib/zip_util.py", line 147, in should_exclude
    return any(fnmatch(fn, p) for p in exclude_patterns)
TypeError: 'NoneType' object is not iterable

percyliang commented:
- Having the bundle show up before uploading is complete is great.  I wonder if it's good to have a new state 'uploading' rather than leaving things as 'created'?  It is easier from the user perspective.
- Note: from the website, however, the bundle doesn't show up until uploading finishes (and there's no progress bar).  I guess that's another PR...

percyliang commented:
LGTM modulo comments (mostly asking for documentation and fast upload)

kashizui commented:
@percyliang: I can get 60+ mbps if I disable on-the-fly compression. So it becomes a tradeoff depending on whether or not the network is a bottleneck for the user. `worksheets-test` for example seems to be capped at 3mbps for upload. Perhaps we should make it an option to the user whether or not to always force compression?

percyliang commented:
@kashizui do you mean 60 MB/sec?

Is this the situation?
localhost: compression is 60, no compression is 6
wstest: compression is ?, no compression is ?
Can't we just check if we're connecting to localhost and switch between the two?

Also, I'm really surprised compression means a factor of 10x - what is going on there?

kashizui commented:
Right, to recap:
localhost: compression is 6 MB/s, **no** compression is 60 MB/s
wstest: compression is 700 KB/s, **no** compression is also 700 KB/s

Check if we're connecting to localhost and/or stanford.edu? (I guess that's also localhost with the ssh tunnel..)

I'll look into why gzip might be 10x slower, definitely doesn't seem right given it's standard on a lot of HTTP services. It may be that we're using the command-line utility that's writing to a pipe, from which we're streaming the compressed data. That's a lot of read and writes that may or may not be avoidable.

kashizui commented:
Using zlib directly, I'm still getting a maximum of 6 MB/s at the default compression level (6). Using the fastest compression level (1) I can get up to 20+ MB/s but not any faster than that. The data doesn't get compressed very much though.

percyliang commented:
> wstest: compression is 700 KB/s, no compression is also 700 KB/s
> It seems like compression is just strictly worse than no compression (on
> wstest it's equal, and on localhost, it's worse).  Is there any reason to
> have it then?

On Fri, Jul 8, 2016 at 12:35 AM, Stephen Koo notifications@github.com
wrote:

> Using zlib directly, I'm still getting a maximum of 6 MB/s at the default
> compression level (6). Using the fastest compression level (1) I can get up
> to 20+ MB/s but not any faster than that. The data doesn't get compressed
> very much though.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codalab/codalab-cli/pull/434#issuecomment-231294894,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAakuKwu4fijQaGx4jW3b1ZbMtH7c4D3ks5qTf25gaJpZM4Ic11h
> .

kashizui commented:
Yes, exactly in the case where the network speed is the bottleneck,
compression is better, because you can end up transferring 60 MB rather
than 160 MB (as was the case for the glove vectors I was uploading as a
test.)
On Fri, Jul 8, 2016 at 08:33 Percy Liang notifications@github.com wrote:

> > wstest: compression is 700 KB/s, no compression is also 700 KB/s
> > It seems like compression is just strictly worse than no compression (on
> > wstest it's equal, and on localhost, it's worse). Is there any reason to
> > have it then?
> 
> On Fri, Jul 8, 2016 at 12:35 AM, Stephen Koo notifications@github.com
> wrote:
> 
> > Using zlib directly, I'm still getting a maximum of 6 MB/s at the default
> > compression level (6). Using the fastest compression level (1) I can get
> > up
> > to 20+ MB/s but not any faster than that. The data doesn't get compressed
> > very much though.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/codalab/codalab-cli/pull/434#issuecomment-231294894
> > ,
> > or mute the thread
> > <
> > https://github.com/notifications/unsubscribe/AAakuKwu4fijQaGx4jW3b1ZbMtH7c4D3ks5qTf25gaJpZM4Ic11h
> > 
> > .
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codalab/codalab-cli/pull/434#issuecomment-231391984,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAlhtTyEWk49tyOQkaHnELYYBBvBjl26ks5qTm3FgaJpZM4Ic11h
> .

percyliang commented:
Hm, I guess minimally it should be an option (can it just be client-side)?

On Fri, Jul 8, 2016 at 10:20 AM, Stephen Koo notifications@github.com
wrote:

> Yes, exactly in the case where the network speed is the bottleneck,
> compression is better, because you can end up transferring 60 MB rather
> than 160 MB (as was the case for the glove vectors I was uploading as a
> test.)
> On Fri, Jul 8, 2016 at 08:33 Percy Liang notifications@github.com wrote:
> 
> > > wstest: compression is 700 KB/s, no compression is also 700 KB/s
> > > It seems like compression is just strictly worse than no compression (on
> > > wstest it's equal, and on localhost, it's worse). Is there any reason to
> > > have it then?
> > 
> > On Fri, Jul 8, 2016 at 12:35 AM, Stephen Koo notifications@github.com
> > wrote:
> > 
> > > Using zlib directly, I'm still getting a maximum of 6 MB/s at the
> > > default
> > > compression level (6). Using the fastest compression level (1) I can
> > > get
> > > up
> > > to 20+ MB/s but not any faster than that. The data doesn't get
> > > compressed
> > > very much though.
> > > 
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> > > https://github.com/codalab/codalab-cli/pull/434#issuecomment-231294894
> > > ,
> > > or mute the thread
> > > <
> > 
> > https://github.com/notifications/unsubscribe/AAakuKwu4fijQaGx4jW3b1ZbMtH7c4D3ks5qTf25gaJpZM4Ic11h
> > 
> > > .
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/codalab/codalab-cli/pull/434#issuecomment-231391984
> > ,
> > or mute the thread
> > <
> > https://github.com/notifications/unsubscribe/AAlhtTyEWk49tyOQkaHnELYYBBvBjl26ks5qTm3FgaJpZM4Ic11h
> > 
> > .
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codalab/codalab-cli/pull/434#issuecomment-231419477,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAakuDRNGaOfPnF_7ygFp85i1ZtxGXz2ks5qTob3gaJpZM4Ic11h
> .

kashizui commented:
Yeah I agree, should minimally be an option. The question is whether compression should be default, or whether there's a reasonable heuristic to choose compression automatically.

Yes fortunately this decision is completely up to the client: the API accepts both compressed and uncompressed and can handle them appropriately.

kashizui committed:
Fix double login

kashizui committed:
Add --force-compression flag to upload

kashizui commented:
@percyliang Addressed all comments. Significant changes:
- Fixed double login issue
- Added --force-compression flag, with no compression on single files as the default, but included note in documentation that using the flag can help with uploads over slow networks.
- Fixed multiple-upload bug and added CLI test for that case.
- Added a new UPLOADING bundle state
- Print progress for downloads as well

kashizui commented:
@percyliang If you think these decisions are OK, we are ready to merge.

percyliang commented:
This is great - thanks!  Let's merge!

On Sat, Jul 9, 2016 at 12:08 AM, Stephen Koo notifications@github.com
wrote:

> @percyliang https://github.com/percyliang If you think these decisions
> are OK, we are ready to merge.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codalab/codalab-cli/pull/434#issuecomment-231520089,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAakuC5B-8_mPRVUEX1_2ElpMEmlG268ks5qT0j5gaJpZM4Ic11h
> .
